File: src/routing.ts
import fs from "node:fs";
import yaml from "js-yaml";
import type { ChatCompletionBody, RoutingPolicyYaml } from "./types.js";
import type { Logger } from "./logger.js";

export class RoutingPolicy {
  private path: string;
  private mtimeMs = 0;
  private ambiguousDefault = "";
  private rules: NonNullable<RoutingPolicyYaml["rules"]> = [];
  private log: Logger;

  constructor(filePath: string, log: Logger) {
    this.path = filePath;
    this.log = log;
  }

  reloadIfStale(): void {
    let st: fs.Stats;
    try {
      st = fs.statSync(this.path);
    } catch (e) {
      this.log.error({ err: e, path: this.path }, "routing policy file missing");
      this.rules = [];
      this.ambiguousDefault = "";
      this.mtimeMs = 0;
      return;
    }
    if (st.mtimeMs === this.mtimeMs) return;
    this.mtimeMs = st.mtimeMs;
    try {
      const raw = fs.readFileSync(this.path, "utf8");
      const doc = yaml.load(raw) as RoutingPolicyYaml | undefined;
      this.ambiguousDefault = doc?.ambiguous_default_model?.trim() ?? "";
      this.rules = doc?.rules ?? [];
      this.log.info(
        { path: this.path, rules: this.rules.length },
        "reloaded routing policy",
      );
    } catch (e) {
      this.log.error(
        { err: e, path: this.path },
        "failed to parse routing policy yaml",
      );
    }
  }

  /** First LiteLLM model id to try for this turn (must appear in fallback chain). */
  pickInitialModel(
    body: ChatCompletionBody,
    fallbackChain: string[],
    virtualModelId: string,
  ): { model: string; via: "rule" | "ambiguous_default" | "chain_only" } {
    this.reloadIfStale();
    if (body.model !== virtualModelId) {
      return { model: body.model ?? "", via: "chain_only" };
    }

    const lastUserChars = lastUserMessageCharCount(body.messages);

    for (const rule of this.rules) {
      if (!rule?.models?.length) continue;
      const when = rule.when ?? {};
      if (
        when.min_message_chars != null &&
        lastUserChars < when.min_message_chars
      ) {
        continue;
      }
      const first = rule.models[0];
      this.log.debug(
        { rule: rule.name ?? "(unnamed)", initialModel: first, lastUserChars },
        "routing rule matched",
      );
      return { model: first, via: "rule" };
    }

    if (this.ambiguousDefault) {
      this.log.debug(
        { initialModel: this.ambiguousDefault, lastUserChars },
        "routing: no rule matched, using ambiguous_default_model",
      );
      return { model: this.ambiguousDefault, via: "ambiguous_default" };
    }

    const first = fallbackChain[0] ?? "";
    this.log.debug(
      { initialModel: first, lastUserChars },
      "routing: no policy default; using first fallback_chain entry",
    );
    return { model: first, via: "chain_only" };
  }
}

function lastUserMessageCharCount(
  messages: ChatCompletionBody["messages"],
): number {
  if (!messages?.length) return 0;
  for (let i = messages.length - 1; i >= 0; i--) {
    const m = messages[i];
    if (m?.role !== "user") continue;
    return contentLength(m.content);
  }
  return 0;
}

function contentLength(c: unknown): number {
  if (typeof c === "string") return c.length;
  if (Array.isArray(c)) {
    let n = 0;
    for (const part of c) {
      if (typeof part === "object" && part && "text" in part) {
        const t = (part as { text?: string }).text;
        if (typeof t === "string") n += t.length;
      }
    }
    return n;
  }
  return 0;
}

/** Index in fallback_chain to start attempts from (must be in chain). */
export function startingFallbackIndex(
  initialModel: string,
  fallbackChain: string[],
): number {
  const idx = fallbackChain.indexOf(initialModel);
  if (idx >= 0) return idx;
  return 0;
}

File: internal/routing/routing.go
package routing

import (
	"encoding/json"
	"log/slog"
	"os"
	"sync"

	"gopkg.in/yaml.v3"
)

// Via matches TypeScript pickInitialModel return.
type Via string

const (
	ViaRule             Via = "rule"
	ViaAmbiguousDefault Via = "ambiguous_default"
	ViaChainOnly        Via = "chain_only"
)

type policyDoc struct {
	AmbiguousDefault string `yaml:"ambiguous_default_model"`
	Rules            []struct {
		Name string `yaml:"name"`
		When struct {
			MinMessageChars *int `yaml:"min_message_chars"`
		} `yaml:"when"`
		Models []string `yaml:"models"`
	} `yaml:"rules"`
}

// Policy reloads routing-policy.yaml on mtime (src/routing.ts).
type Policy struct {
	path             string
	log              *slog.Logger
	mu               sync.Mutex
	mtimeNs          int64
	ambiguousDefault string
	rules            []policyRule
}

type policyRule struct {
	name            string
	minMessageChars *int
	models          []string
}

func NewPolicy(path string, log *slog.Logger) *Policy {
	return &Policy{path: path, log: log}
}

func (p *Policy) ReloadIfStale() {
	p.mu.Lock()
	defer p.mu.Unlock()

	st, err := os.Stat(p.path)
	if err != nil {
		if p.log != nil {
			p.log.Error("routing policy file missing", "path", p.path, "err", err)
		}
		p.rules = nil
		p.ambiguousDefault = ""
		p.mtimeNs = 0
		return
	}
	mt := st.ModTime().UnixNano()
	if mt == p.mtimeNs {
		return
	}
	p.mtimeNs = mt

	raw, err := os.ReadFile(p.path)
	if err != nil {
		if p.log != nil {
			p.log.Error("read routing policy", "path", p.path, "err", err)
		}
		return
	}
	var doc policyDoc
	if err := yaml.Unmarshal(raw, &doc); err != nil {
		if p.log != nil {
			p.log.Error("failed to parse routing policy yaml", "path", p.path, "err", err)
		}
		return
	}
	p.ambiguousDefault = doc.AmbiguousDefault
	var rules []policyRule
	for _, r := range doc.Rules {
		if len(r.Models) == 0 {
			continue
		}
		rules = append(rules, policyRule{
			name:            r.Name,
			minMessageChars: r.When.MinMessageChars,
			models:          r.Models,
		})
	}
	p.rules = rules
	if p.log != nil {
		p.log.Info("reloaded routing policy", "path", p.path, "rules", len(p.rules))
	}
}

// PickInitialModel returns the first upstream model id for the virtual Claudia model.
func (p *Policy) PickInitialModel(body map[string]json.RawMessage, fallbackChain []string, virtualModelID string) (model string, via Via) {
	p.ReloadIfStale()

	var clientModel string
	if m, ok := body["model"]; ok {
		_ = json.Unmarshal(m, &clientModel)
	}
	if clientModel != virtualModelID {
		return clientModel, ViaChainOnly
	}

	lastUser := lastUserMessageCharCount(body)

	p.mu.Lock()
	defer p.mu.Unlock()

	for _, rule := range p.rules {
		if len(rule.models) == 0 {
			continue
		}
		if rule.minMessageChars != nil && lastUser < *rule.minMessageChars {
			continue
		}
		if p.log != nil {
			p.log.Debug("routing rule matched", "rule", rule.name, "initialModel", rule.models[0], "lastUserChars", lastUser)
		}
		return rule.models[0], ViaRule
	}

	if p.ambiguousDefault != "" {
		if p.log != nil {
			p.log.Debug("routing: no rule matched, using ambiguous_default_model", "initialModel", p.ambiguousDefault, "lastUserChars", lastUser)
		}
		return p.ambiguousDefault, ViaAmbiguousDefault
	}

	first := ""
	if len(fallbackChain) > 0 {
		first = fallbackChain[0]
	}
	if p.log != nil {
		p.log.Debug("routing: no policy default; using first fallback_chain entry", "initialModel", first, "lastUserChars", lastUser)
	}
	return first, ViaChainOnly
}

func lastUserMessageCharCount(body map[string]json.RawMessage) int {
	raw, ok := body["messages"]
	if !ok {
		return 0
	}
	var messages []struct {
		Role    string          `json:"role"`
		Content json.RawMessage `json:"content"`
	}
	if err := json.Unmarshal(raw, &messages); err != nil {
		return 0
	}
	for i := len(messages) - 1; i >= 0; i-- {
		if messages[i].Role != "user" {
			continue
		}
		return contentLength(messages[i].Content)
	}
	return 0
}

func contentLength(raw json.RawMessage) int {
	if len(raw) == 0 {
		return 0
	}
	if raw[0] == '"' {
		var s string
		if err := json.Unmarshal(raw, &s); err != nil {
			return 0
		}
		return len([]rune(s))
	}
	if raw[0] == '[' {
		var parts []struct {
			Text string `json:"text"`
		}
		if err := json.Unmarshal(raw, &parts); err != nil {
			return 0
		}
		n := 0
		for _, part := range parts {
			n += len([]rune(part.Text))
		}
		return n
	}
	return 0
}

// StartingFallbackIndex is src/routing.ts startingFallbackIndex.
func StartingFallbackIndex(initialModel string, fallbackChain []string) int {
	for i, m := range fallbackChain {
		if m == initialModel {
			return i
		}
	}
	return 0
}
